### PR TITLE
Add unit tests for Identity core

### PR DESCRIPTION
**Pull request by pulasthi7**
### Proposed changes in this pull request

Unit tests for Identity core

### When should this PR be merged

After https://github.com/wso2/carbon-identity-framework/pull/1114

### Follow up actions

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+
+    @Test(dataProvider = "getIdentityErrorMsgTestData")
+    public void testIdentityErrorMsg(String code, int failedAttempts, int maxAttempts) throws Exception {
+
+        IdentityErrorMsgContext msgContext = new IdentityErrorMsgContext(code, failedAttempts, maxAttempts);
+        assertNull(IdentityUtil.getIdentityErrorMsg(), "Error msg should be null initially");
+        IdentityUtil.setIdentityErrorMsg(msgContext);
+        assertEquals(IdentityUtil.getIdentityErrorMsg().getErrorCode(), code);
+        assertEquals(IdentityUtil.getIdentityErrorMsg().getFailedLoginAttempts(), failedAttempts);
+        assertEquals(IdentityUtil.getIdentityErrorMsg().getMaximumLoginAttempts(), maxAttempts);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: Shall we put error messages for these assertEquals?

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+        };
+    }
+
+    @Test(dataProvider = "getPropertyTestData")
+    public void testGetProperty(String key, Object value, String expected) throws Exception {
+        Map<String, Object> mockConfig = new HashMap<>();
+        mockConfig.put(key, value);
+
+        Whitebox.setInternalState(IdentityUtil.class, "configuration", mockConfig);
+        assertEquals(IdentityUtil.getProperty(key), expected);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: Shall we add error message?

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+
+        IdentityEventListenerConfigKey key = new IdentityEventListenerConfigKey("ListenerType", "ListenerName");
+        IdentityEventListenerConfig config = new IdentityEventListenerConfig("true", 0, key, new Properties());
+        Map<IdentityEventListenerConfigKey, IdentityEventListenerConfig> mockedMap = new HashMap<>();
+        mockedMap.put(key, config);
+        Whitebox.setInternalState(IdentityUtil.class, "eventListenerConfiguration", mockedMap);
+
+        IdentityEventListenerConfig configResponse = IdentityUtil.readEventListenerProperty("ListenerType", "ListenerName");
+        assertEquals(configResponse.getEnable(), "true");
+        assertEquals(configResponse.getOrder(), 0);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error messages.

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+
+        IdentityEventListenerConfig nullResponse = IdentityUtil.readEventListenerProperty("NonExistingType",
+                "NonExistingListenerName");
+        assertNull(nullResponse);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error message

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,761 @@
...
+                {"manager1", "", null},
+                {"manager2", "", cacheConfig2},
+                {"manager1", "$__local__$.key1", cacheConfig1},
+        };
+    }
+
+    @Test(dataProvider = "getIdentityCacheConfigTestData")
+    public void testGetIdentityCacheConfig(String cacheManagerName, String cacheName, Object expected)
+            throws Exception {
+        assertEquals(IdentityUtil.getIdentityCacheConfig(cacheManagerName, cacheName), expected);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error message

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,761 @@
...
+    }
+
+    @Test
+    public void testGetIdentityCookiesConfigurationHolder() throws Exception {
+
+        Map<String, IdentityCookieConfig> mockIdentityCookiesConfigurationHolder = new HashMap<>();
+        mockIdentityCookiesConfigurationHolder.put("cookie", new IdentityCookieConfig("cookieName"));
+        Whitebox.setInternalState(IdentityUtil.class, "identityCookiesConfigurationHolder",
+                mockIdentityCookiesConfigurationHolder);
+        assertEquals(IdentityUtil.getIdentityCookiesConfigurationHolder(), mockIdentityCookiesConfigurationHolder);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error message

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,761 @@
...
+
+    @Test
+    public void testGetIdentityCookieConfig() throws Exception {
+        Map<String, IdentityCookieConfig> mockIdentityCookiesConfigurationHolder = new HashMap<>();
+        IdentityCookieConfig cookieConfig = new IdentityCookieConfig("cookieName");
+        mockIdentityCookiesConfigurationHolder.put("cookie", cookieConfig);
+        Whitebox.setInternalState(IdentityUtil.class, "identityCookiesConfigurationHolder",
+                mockIdentityCookiesConfigurationHolder);
+        assertEquals(IdentityUtil.getIdentityCookieConfig("cookie"), cookieConfig);
+        assertNull(IdentityUtil.getIdentityCookieConfig("nonExisting"));
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error message

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,761 @@
...
+        when(IdentityConfigParser.getIdentityCacheConfigurationHolder()).thenReturn(mockedCacheConfig);
+        when(IdentityConfigParser.getIdentityCookieConfigurationHolder()).thenReturn(mockedCookieConfig);
+        when(IdentityConfigParser.getInstance()).thenReturn(mockConfigParser);
+        IdentityUtil.populateProperties();
+        assertEquals(Whitebox.getField(IdentityUtil.class, "configuration").get(IdentityUtil.class), mockConfig);
+        assertEquals(Whitebox.getField(IdentityUtil.class, "eventListenerConfiguration").get(IdentityUtil.class),
+                mockedEventListenerConfig);
+        assertEquals(IdentityUtil.getIdentityCookiesConfigurationHolder(), mockedCookieConfig);
+        assertEquals(Whitebox.getField(IdentityUtil.class, "identityCacheConfigurationHolder").get(IdentityUtil
+                .class), mockedCacheConfig);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error messages

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,761 @@
...
+    public Object[][] getPPIDDisplayValueTestData() {
+        return new Object[][]{
+                {"aabfcdfe", "26G-NDT6-656"},
+                {"", "UT3-EYBB-BDJ"},
+        };
+    }
+
+    @Test(dataProvider = "getPPIDDisplayValueTestData")
+    public void testGetPPIDDisplayValue(String provided, String expected) throws Exception {
+        assertEquals(IdentityUtil.getPPIDDisplayValue(provided), expected);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error message

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 10]
```diff
@@ -0,0 +1,761 @@
...
+    @Test
+    public void testNodeToString() throws Exception {
+        String xmlText = "<TestNode attr=\"attrValue\">Test value</TestNode>";
+
+        OMDOMFactory fac = new OMDOMFactory();
+        Document document = (Document) fac.createOMDocument();
+        Element testNode = document.createElement("TestNode");
+        testNode.setTextContent("Test value");
+        testNode.setAttribute("attr", "attrValue");
+        assertEquals(IdentityUtil.nodeToString(testNode), xmlText);
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: error message

File: components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+
+<suite name="identity-core-test-suite">
+    <test name="identity-core-test-all">
+        <classes>
+            <class name="org.wso2.carbon.identity.core.util.IdentityUtilTest"/>
+        </classes>
+    </test>
+</suite>
```
Comment (thanujalk) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml]: add a new line.

File: components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java [hunk 12]
```diff
@@ -615,7 +615,7 @@ public static String extractDomainFromName(String nameWithDomain) {
      * @return application name with domain name
      */
     public static String addDomainToName(String name, String domainName) {
-        if (domainName != null && !StringUtils.contains(name, UserCoreConstants.DOMAIN_SEPARATOR)) {
+        if (domainName != null && name!=null && !name.contains(UserCoreConstants.DOMAIN_SEPARATOR)) {
```
Comment (mefarazath) [inline components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/util/IdentityUtil.java]: Check formatiing!

File: components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java [hunk 13]
```diff
@@ -0,0 +1,807 @@
...
+        assertFalse(IdentityUtil.isSupportedByUserStore(mockUserStoreManager, "op2"), "Expected false for op2");
+        assertTrue(IdentityUtil.isSupportedByUserStore(mockUserStoreManager, "op3"), "Expected true for op3");
+        assertTrue(IdentityUtil.isSupportedByUserStore(null, "op4"), "Expected true for op4 in null userstore");
+    }
+
+    @ObjectFactory
+    public IObjectFactory getObjectFactory() {
+        return new org.powermock.modules.testng.PowerMockObjectFactory();
+    }
+}
```
Comment (mefarazath) [inline components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: add a new line at the end of file

Comment (pulasthi7) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/util/IdentityUtilTest.java]: Done

File: components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml (hunk 11 above)
Comment (pulasthi7) [reply components/identity-core/org.wso2.carbon.identity.core/src/test/resources/testng.xml]: Done

**Pull request merged**